dbms/src/Interpreters/UsersManager.cpp

```diff
@@ -63,7 +63,7 @@ bool UsersManager::hasAccessToDatabase(const std::string & user_name, const std:
         throw Exception("Unknown user " + user_name, ErrorCodes::UNKNOWN_USER);
 
     auto user = it->second;
-    return user->databases.empty() || user->databases.count(database_name);
+    return !user->has_allowed_databases || user->databases.count(database_name);
```

Comment:
I suggest to use `std::optional` to make the code more clear.

Reply:
Changed to std::optional